_posts/2021-10-29-swift-interface-for-node-js.md

```diff
@@ -0,0 +1,179 @@
+she goes on stage to reveal that she just wants a normal Christmas instead of
+having to celebrate two Christmases in two families[^1].
+
+Yes, I sometimes want my 2 co-languages to live under the same roof.
+
+## How can JavaScript & Swift interact
+
+JavaScript and Swift can interact in many different ways. A non-exhaustive list:
+
+- **Running JavaScript from Swift code**. Thanks to [JavaScriptCore][6], this is
+  actually pretty easy. I gave [a talk about it][8] during a Paris CocoaHeads
+  session.[^2];
+- **Running Swift code from JavaScript inside the browser**. Thanks to the
+  [Swift WASM project][9], this is also actually pretty easy. But I couldn’t
+  find a good use case for it. And during my POC, a simple Hello World being
+  displayed in the browser would weigh almost 40 MB so I didn’t pursue this
+  lead.
+
+{% asset wasm-hello-world.png alt="A screenshot of a WASM hello world weighing 38.47 MB" %}
+
+- **👉 Running Swift code from JavaScript in the terminal with Node 👈**. This
+  is actually what this post is about! A use case could be to write a command
+  line interface that would use [oclif][7] for user interactions while running
+  Swift code under the hood. Sweet.
+- I am sure there are many other possible interactions. If you can think of one,
+  pull requests are welcome.
+
+A classic way for 2 languages to run hand-in-hand is actually to run
+hand-in-hand-in-hand with a dynamic library in the middle. So our plan is pretty
+simple:
+
+- 1️⃣ Write some Swift code;
+- 2️⃣ Turn this code into a dynamic library;
+- 3️⃣ Use this dynamic library with Node.
+
+## Writing a dynamic library in Swift
+
+In real-life, the plan was trickier and there was a lot of back and forth
+between all of the steps.
+
+Using Swift Package Manager, creating a dynamic library is done like so:
+
+```swift
+// Package.swift
+.library(name: "MyDynamicLibrary", type: .dynamic, targets: ["MyLibrary"])
+```
+
+Pretty basic. Let’s move on to a simple function that will greet whatever will
+passed as an argument.
+
+My oh-so-naive first draft of interfacable code looked like this:
+
+```swift
+public func greet(name: String) -> String {
+    return "Hello, \(name)!"
+}
+```
+
+But my successful attempt ended up looking like this:
+
+```swift
+@_cdecl("greet")
+public func greet(
+    cStringName: UnsafePointer<Int8>,
+    outputSize: Int16,
+    output: UnsafeMutablePointer<Int8>
+) -> Bool {
+    let name = String(cString: cStringName)
+    let cString = "Hello, \(name)!".cString(using: .utf8)!
+    for (index, char) in cString.enumerated() {
+        if (index >= outputSize) {
+            return false
+        }
+        output.advanced(by: index).pointee = char
+    }
+    return true
+}
+```
+
+Lessons learned:
+
+- An interoperable string is actually called a C String, which translates to a
+  `UnsafePointer<Int8>`;
+- I would get an error returning the output as a result of the function. My
```

Comment:
what is the lesson here? the last line? if so, start with that, and then explain how you got there.